lightningd/jsonrpc.c

```diff
@@ -237,48 +237,58 @@ static const struct json_command *find_cmd(const char *buffer,
 	return NULL;
 }
 
-static void connection_result(struct json_connection *jcon,
-			      const char *id,
-			      const char *res,
-			      const char *err,
-			      int code,
-			      const struct json_result *data)
+static void json_done(struct json_connection *jcon, const char *json TAKES)
 {
 	struct json_output *out = tal(jcon, struct json_output);
-	const tal_t *tmpctx;
-	const char* data_str;
-
-	if (err == NULL)
-		out->json = tal_fmt(out,
-				    "{ \"jsonrpc\": \"2.0\", "
-				    "\"result\" : %s,"
-				    " \"id\" : %s }\n",
-				    res, id);
-	else {
-		tmpctx = tal_tmpctx(out);
-		if (data)
-			data_str = tal_fmt(tmpctx,
-					   ", \"data\" : %s",
-					   json_result_string(data));
-		else
-			data_str = "";
-
-		out->json = tal_fmt(out,
-				    "{ \"jsonrpc\": \"2.0\", "
-				    " \"error\" : "
-				      "{ \"code\" : %d,"
-				      " \"message\" : %s%s },"
-				    " \"id\" : %s }\n",
-				    code, err, data_str,
-				    id);
-		tal_free(tmpctx);
-	}
+	out->json = tal_strdup(out, json);
 
 	/* Queue for writing, and wake writer (and maybe reader). */
 	list_add_tail(&jcon->output, &out->list);
 	io_wake(jcon);
 }
 
+static void connection_complete_ok(struct json_connection *jcon,
+				   const char *id,
+				   const struct json_result *result)
+{
+	/* This JSON is simple enough that we build manually */
+	json_done(jcon, take(tal_fmt(jcon,
+				     "{ \"jsonrpc\": \"2.0\", "
+				     "\"result\" : %s,"
+				     " \"id\" : %s }\n",
+				     json_result_string(result), id)));
+}
+
+static void connection_complete_error(struct json_connection *jcon,
+				      const char *id,
+				      const char *errmsg,
+				      int code,
+				      const struct json_result *data)
+{
+	/* Use this to escape errmsg. */
+	struct json_result *errorres = new_json_result(jcon);
+	const char *data_str;
+
+	json_add_string_escape(errorres, NULL, errmsg);
+	if (data)
```

Comment:
The code here somewhat misleadingly makes it appear that `data_str` is somehow related to `errorres`: they are in the same "paragraph" of code, and `data_str` is allocated with `errorres` as parent.

Maybe "cleaner" with a `tal_tmpctx` locally that parents both `errorres` and `data_str`.